### ISSUE TITLE
pass macro not parsed in substituted attribute value

### ISSUE THREAD

**Issue by elextr** [asciidoctor/asciidoctor#2630]
Asciidoctor 1.5.4 (yes, but its all thats in the Ubuntu LTS distro)

Example:
```
= foo

==blah

:abc: foo pass:[blah] foo

blah foo pass:[blah] foo blah pass correctly parsed

[subs=-macros]
blah foo pass:[blah] foo blah pass correctly not parsed

blah {abc} blah pass incorrectly not parsed

[subs="attributes,macros"]
blah {abc} blah pass still incorrectly not parsed even if macros is after attribute subs

[subs=-macros]
blah {abc} blah pass correctly not parsed
```

**Comment by mojavelinux** [comment]
This is by design. Asciidoctor only supports putting the pass macro around the whole value. You'd need to use a pass around the value, then use a pass in the value.

**Comment by elextr** [comment]
To have the macro in the attribute value parsed the whole value seems to have to be in a `pass:m[]`.  But then the explicit specification of the macro substitution overrules the `subs=` setting on the last paragraph in the example above.

What was the background to the decision of only allowing the whole value to be wrapped in pass?

**Comment by mojavelinux** [comment]
This is just not an important problem to solve right now. We have much bigger and pressing issues. I have no intention of changing this behavior right now in light of that.

**Comment by elextr** [comment]
> This is just not an important problem to solve right now. We have much bigger and pressing issues. I have no intention of changing this behavior right now in light of that.

Fine, its your project, but you could still share the background to the decision.

**Comment by mojavelinux** [comment]
Every attribute value has an implicit pass:s,a[] macro around it unless you wrap it in a different pass macro. It's my understanding that's how AsciiDoc has always worked and why I implemented it this way.

```
:abc: foo pass:[blah] foo

[subs=-macros]
blah foo pass:[blah] foo blah pass correctly not parsed
```

Inline macros are not substituted by default in an attribute value and you've explicitly disabled this substitution in the block. Why would you expect macros to be substituted then?

> its your project

I'm not saying this is my project and that's how it goes. As the leader of it, I'm responding to the 1,000 of requests (you could even say shouts and screams) for much more important issues. We just need to focus or we're never going to move forward. Trying to get into fine-grained edge cases before we solve the pressing issues works against the forward progress.

**Comment by mojavelinux** [comment]
As you can probably tell, there's an insane amount of pressure on me right now.

**Comment by elextr** [comment]
> Inline macros are not substituted by default in an attribute value and you've explicitly disabled this substitution in the block. Why would you expect macros to be substituted then?

The attribute value needs to be wrapped in a `pass:m[]` to make the `pass:[]` inside the attribute value get substituted, the `subs=macros` or `subs=-macros` from the block has no effect on the substituted attribute, even if its not wrapped in `pass`, or if its wrapped in `pass:[]`, or (as expected) if its wrapped in `pass:m[]`.

> As the leader of it, I'm responding to the 1,000 of requests (you could even say shouts and screams) for much more important issues. 

> As you can probably tell, there's an insane amount of pressure on me right now.

I'm old enough and grumpy enough to understand, and no offence is taken, and I might add none is ever intended when I myself sound grumpy :)

All I'm trying to determine is, is this corner case:

1. part of the Asciidoc language
2. an implementation specific decision in Asciidoc Python or Asciidoctor (in which case why was it decided, maybe it _should_ be part of the language, or maybe it was just a limitation at the time)
3. a bug in the specific implementation

If its the latter, happy to have alerted you to it, the priority for fixing (if at all) is entirely the Asciidoctor project's decision.

If its one of the first two I'm more interested, and it would seem at the moment to be a bit of both 2 and 3 but definitely not 1 ie part of the language, so I'll proceed on that basis.

**Comment by mojavelinux** [comment]
Oops, I misread your example.

I now understand what you are pointing out.

You are correct, this is a difference. But what confused me is that this is not about attribute value substitution at all. It has to do with how Asciidoctor applies substitutions on the value of an attribute reference. For performance reasons, Asciidoctor does not recheck for pass macros after attribute values are substituted (since passthroughs are handled first). But it's more than just performance. It has to do with how passthroughs are stacked and when they are available. And altering that behavior would require a major change in how substitutions are done (I think).

In other words, passthroughs are not like normal macros. They get extracted first and replaced last, no matter where macros are listed in the substitution order.

**Comment by mojavelinux** [comment]
I do apologize for sounding grumpy and snapping at you. I'm just seeing the issues pile up and not a lot of focus to get the ones we already have resolved. That's just a general sentiment.

**Comment by mojavelinux** [comment]
My feeling about the passthrough has always been that it's really just a hack (albeit a necessary one right now). Until we actually address the underlying problem with it, which is to parse recursively, all we do is just move the hack from one place to another.

**Comment by elextr** [comment]
>  But what confused me is that this is not about attribute value substitution at all.

Correct, its about processing post substitution, but "substitution into the attribute value after its substituted" was such a mouthful to repeat several times that I took a linguistic shortcut, sorry for the confusion.

>  Until we actually address the underlying problem with it, which is to parse recursively, all we do is just move the hack from one place to another.

Yep, I'm parsing it recursively, so lets see how it works out for me.

**Comment by mojavelinux** [comment]
> Yep, I'm parsing it recursively, so lets see how it works out for me.

Excellent! I'll be interested to hear how it goes and what you find.
